cranelift/assembler-x64/meta/src/generate/format.rs

```diff
@@ -159,97 +194,71 @@ impl dsl::Format {
+        };
 
         fmtln!(f, "vex.encode(buf);");
+        style
     }
 
-    fn generate_modrm_byte(
-        &self,
-        f: &mut Formatter,
-        opcode_mod: Option<dsl::OpcodeMod>,
-        digit: Option<dsl::ModRmKind>,
-    ) {
-        use dsl::OperandKind::{FixedReg, Imm, Mem, Reg, RegMem};
-
-        // Some instructions will never emit a ModR/M byte.
+    fn generate_modrm_byte(&self, f: &mut Formatter, modrm_style: ModRmStyle) {
         let operands = self.operands_by_kind();
-        if opcode_mod.is_some()
-            || matches!(
-                operands.as_slice(),
-                [FixedReg(_)] | [FixedReg(_), FixedReg(_)] | [FixedReg(_), Imm(_)]
-            )
-        {
-            f.empty_line();
-            f.comment("No need to emit a ModRM byte.");
-            return;
-        }
-
-        // If we must, emit the ModR/M byte and the SIB byte (if necessary).
-        f.empty_line();
-        f.comment("Emit ModR/M byte.");
         let bytes_at_end = match operands.as_slice() {
-            [.., Imm(imm)] => imm.bytes(),
+            [.., dsl::OperandKind::Imm(imm)] => imm.bytes(),
             _ => 0,
         };
-        match operands.as_slice() {
-            [Reg(reg), Imm(_)] => {
-                let digit = digit.unwrap().unwrap_digit();
+
+        f.empty_line();
+
+        match modrm_style {
+            ModRmStyle::None => f.comment("No need to emit a ModRM byte."),
+            _ => f.comment("Emit ModR/M byte."),
+        }
+
+        match modrm_style {
+            ModRmStyle::None => {}
+            ModRmStyle::RegWithDigit { reg, digit } => {
                 fmtln!(f, "let digit = 0x{digit:x};");
                 fmtln!(f, "self.{reg}.encode_modrm(buf, digit);");
             }
-            [Mem(mem), Imm(_)]
-            | [RegMem(mem), Imm(_)]
-            | [RegMem(mem)]
-            | [FixedReg(_), RegMem(mem)]
-            | [RegMem(mem), FixedReg(_)]
-            | [FixedReg(_), FixedReg(_), RegMem(mem)] => {
-                let digit = digit.unwrap().unwrap_digit();
+            ModRmStyle::RegMemWithDigit { mem, digit } => {
                 fmtln!(f, "let digit = 0x{digit:x};");
                 fmtln!(
                     f,
                     "self.{mem}.encode_rex_suffixes(buf, off, digit, {bytes_at_end});"
                 );
             }
-            [Reg(reg), RegMem(mem)]
-            | [Reg(reg), RegMem(mem), Imm(_)]
-            | [Mem(mem), Reg(reg)]
-            | [Reg(reg), Reg(_), RegMem(mem)]
-            | [RegMem(mem), Reg(reg)]
-            | [RegMem(mem), Reg(reg), Imm(_)]
-            | [RegMem(mem), Reg(reg), FixedReg(_)]
-            | [Reg(reg), Reg(_), RegMem(mem), Imm(_)]
-            | [Reg(reg), Mem(mem)] => {
+            ModRmStyle::RegMem { reg, mem } => {
                 fmtln!(f, "let reg = self.{reg}.enc();");
                 fmtln!(
                     f,
                     "self.{mem}.encode_rex_suffixes(buf, off, reg, {bytes_at_end});"
```

Comment:
Ah, I see: I guess we do need to differentiate `RegMem` and `RegReg` for this kind of thing.